src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java

```diff
@@ -339,25 +341,28 @@ void distributeFlushTasks() {
 
     while (itr.hasNext()) {
       List<List<ChannelData<T>>> blobData = new ArrayList<>();
-      float totalBufferSize = 0;
+      AtomicReference<Float> totalBufferSize = new AtomicReference<>((float) 0);
 
       final String filePath = getFilePath(this.targetStage.getClientPrefix());
 
       // Distribute work at table level, create a new blob if reaching the blob size limit
-      while (itr.hasNext() && totalBufferSize <= MAX_BLOB_SIZE_IN_BYTES) {
+      while (itr.hasNext() && totalBufferSize.get() <= MAX_BLOB_SIZE_IN_BYTES) {
         ConcurrentHashMap<String, SnowflakeStreamingIngestChannelInternal<T>> table =
             itr.next().getValue();
-        List<ChannelData<T>> channelsDataPerTable = new ArrayList<>();
-        // TODO: we could do parallel stream to get the channelData if needed
-        for (SnowflakeStreamingIngestChannelInternal<T> channel : table.values()) {
-          if (channel.isValid()) {
-            ChannelData<T> data = channel.getData(filePath);
-            if (data != null) {
-              channelsDataPerTable.add(data);
-              totalBufferSize += data.getBufferSize();
-            }
-          }
-        }
+        List<ChannelData<T>> channelsDataPerTable = Collections.synchronizedList(new ArrayList<>());
+        // Use parallel stream since getData could be the performance bottleneck when the number of
+        // channels are big
```

Comment:
Will this cause performance when the number of channels is small to degrade?

Reply:
Not really based on my quick testing, the overhead of using a parallel stream is negligible compared with the time to get the buffer data

Reply:
Ack